### PR TITLE
Security/permissions

### PR DESCRIPTION
**Pull request by Vassyli**
Adds permissions to the crate and an authorization service. The latter provides methods to deal with authorization, such as checking login status, or check if the current user has a specific superuser flag set. A TypeGuardian can be used to wrap Type classes to make sure only specific fields can be accessed. This allowes the types to be as agnostic to authorization as possible and centralizes it within the resolvers.

Existing resolvers were updated properly check authorizations. Tests were changed, too.

### CODE REVIEW

File: src/Services/CoreGameService.php [hunk 1]
```diff
@@ -80,4 +80,8 @@ public function getModuleManager(): ModuleManager
     {
         return $this->game->getModuleManager();
     }
+
+    public function test() {
```
Comment (austenmc) [inline src/Services/CoreGameService.php]: What is this?

File: tests/WebTestCase.php [hunk 2]
```diff
@@ -137,21 +137,25 @@ protected function assertJsonContainsKeys($expectedKeys, $client)
         }
     }
 
-
-    protected function queryHelper($query, $jsonVariables)
+    /*protected function queryHelper($query, $jsonVariables, $apiKey = null)
```
Comment (austenmc) [inline tests/WebTestCase.php]: Reason to keep the commented code?

File: src/Models/UserPermissionAssociation.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
+<?php
+/**
```
Comment (austenmc) [inline src/Models/UserPermissionAssociation.php]: nit: let's remove these kinds of comments

File: src/Services/AuthorizationService.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        $user = $this->container->get('security.token_storage')->getToken()->getUser();
+
+        if ($user instanceof User) {
+            return $user;
+        } else {
+            return null;
+        }
+    }
+
+    public function isUser($object) {
```
Comment (austenmc) [inline src/Services/AuthorizationService.php]: This is more like `isCurrentUser` right? Not just if the object passed in is a user.

File: src/AppBundle/GraphQL/Types/TypeGuardian.php [hunk 5]
```diff
@@ -0,0 +1,57 @@
+<?php
+/**
```
Comment (austenmc) [inline src/AppBundle/GraphQL/Types/TypeGuardian.php]: nit: remove comment

File: src/AppBundle/GraphQL/Types/TypeGuardian.php [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+ * Created by PhpStorm.
+ * User: sauterb
+ * Date: 15/03/17
+ * Time: 16:34
+ */
+
+namespace LotGD\Crate\GraphQL\AppBundle\GraphQL\Types;
+
+
+class TypeGuardian implements TypeInterface
```
Comment (austenmc) [inline src/AppBundle/GraphQL/Types/TypeGuardian.php:7]: This is creative :)

File: src/Services/AuthorizationService.php [hunk 7]
```diff
@@ -41,29 +49,15 @@ public function getCurrentUser(): ?User
         return $this->user;
     }
 
+    /**
+     * Short method to check if a user is logged in or not.
+     * @return bool
+     */
     public function isLoggedin(): bool
```
Comment (austenmc) [inline src/Services/AuthorizationService.php]: nit: `isLoggedIn`

File: tests/Functional/GraphQL/UserQueryTest.php [hunk 8]
```diff
@@ -51,13 +51,11 @@ public function testIfKnownUserCannotBeRetrievedByNameWithoutProperAuthorization
     "data": {
         "user":null
     },
-    "extensions": {
-        "warnings": [{
-            "message": "Access denied to this field.",
+    "errors": [{
+        "message": "Accessing this field with this parameters is not allowed.",
```
Comment (austenmc) [inline tests/Functional/GraphQL/UserQueryTest.php]: nit: "these parameters"

File: tests/Functional/GraphQL/UserQueryTest.php [hunk 9]
```diff
@@ -117,13 +115,11 @@ public function testIfAnUnknownUserRetrievedByNameReturnsStillAnAccessError()
...
     },
-    "extensions": {
-        "warnings": [{
-            "message": "Access denied to this field.",
-            "locations": [{"line":2,"column":5}],
-            "path": ["user"]
-        }]
-    }
+    "errors": [{
+        "message": "Accessing this field with this parameters is not allowed.",
```
Comment (austenmc) [inline tests/Functional/GraphQL/UserQueryTest.php]: nit: "these parameters"

**Review by austenmc — APPROVED** [review]

File: tests/WebTestCase.php (hunk 2 above)
Comment (Vassyli) [reply tests/WebTestCase.php]: None, oversight by me.

File: src/Models/UserPermissionAssociation.php (hunk 3 above)
Comment (Vassyli) [reply src/Models/UserPermissionAssociation.php]: Thanks, overlooked that one.

File: src/AppBundle/GraphQL/Types/TypeGuardian.php (hunk 6 above)
Comment (Vassyli) [reply src/AppBundle/GraphQL/Types/TypeGuardian.php:7]: Thank you :)

File: src/Services/AuthorizationService.php (hunk 4 above)
Comment (Vassyli) [reply src/Services/AuthorizationService.php]: Didn't I completely remove that part? Wording was temporary anyway, but you're right.

File: src/Services/CoreGameService.php (hunk 1 above)
Comment (Vassyli) [reply src/Services/CoreGameService.php]: Something that's no longer in there. I used this to test if the graphql-scheme "access" field correctly access that method. Removed it and changes the way to control access to fields anyway.

File: src/AppBundle/GraphQL/Types/TypeGuardian.php (hunk 5 above)
Comment (Vassyli) [reply src/AppBundle/GraphQL/Types/TypeGuardian.php]: dito.

**Pull request merged**